application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -0,0 +1,120 @@
+ */
+public final class WarnCommand extends SlashCommandAdapter {
+    private static final String USER_OPTION = "user";
+    private static final String REASON_OPTION = "reason";
+    private static final String ACTION_VERB = "warn";
+    private final ModerationActionsStore actionsStore;
+    private final Predicate<String> hasRequiredRole;
+
+    /**
+     * Creates a new Instance.
+     *
+     * @param actionsStore used to store actions issued by this command
+     */
+    public WarnCommand(@NotNull ModerationActionsStore actionsStore) {
+        super("warn", "warns the user", SlashCommandVisibility.GUILD);
+
+        getData().addOption(OptionType.USER, USER_OPTION, "The user to warn", true)
+            .addOption(OptionType.STRING, REASON_OPTION, "The reason for the warning", true);
+
+        hasRequiredRole = Pattern.compile(Config.getInstance().getHeavyModerationRolePattern())
+            .asMatchPredicate();
+        this.actionsStore = Objects.requireNonNull(actionsStore);
+    }
+
+    private static @NotNull RestAction<Boolean> dmUser(long userId, @NotNull String reason,
+            @NotNull SlashCommandEvent event) {
+        return event.getJDA()
+            .openPrivateChannelById(userId)
+            .flatMap(channel -> channel.sendMessage(
+                    """
+                            Hey there, We are sorry to inform you that you have been warned for the following reason: %s
+                            The repercussion of this warning depends on the severity of the warning or one the number of warnings you have received.\040
+                            The warn system can be seen in the <#652440333107331082> channel.\040
+                            """
+                        .formatted(reason)))
+            .mapToResult()
+            .map(Result::isSuccess);
+    }
+
+    private static @NotNull MessageEmbed sendFeedback(boolean hasSentDm, @NotNull User target,
+            @NotNull Member author, @NotNull String reason) {
+        String dmNoticeText = "";
+        if (!hasSentDm) {
+            dmNoticeText = "(Unable to send them a DM.)";
+        }
+        return ModerationUtils.createActionResponse(author.getUser(), ModerationUtils.Action.WARN,
+                target, dmNoticeText, reason);
+    }
+
+    private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
+            @NotNull CharSequence reason, @NotNull Guild guild, @NotNull Interaction event) {
+
+        if (!ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event)) {
+            return false;
+        }
+        if (!ModerationUtils.handleHasBotPermissions(ACTION_VERB, Permission.BAN_MEMBERS, bot,
+                guild, event)) {
+            return false;
+        }
+        if (!ModerationUtils.handleHasAuthorPermissions(ACTION_VERB, Permission.BAN_MEMBERS, author,
+                guild, event)) {
+            return false;
+        }
```

Comment:
I would remove those checks. I do not think there is any benefit in tying `/warn` to `BAN_MEMBERS`. Just the `hasRequiredRole` should be enough. (similar to how `/audit` works).